### ISSUE TITLE
Multi server file storage

### ISSUE THREAD

**Issue by mattsbanner** [putyourlightson/craft-blitz#337]
### Is your feature request related to a problem? Please describe.
I have load balanced servers, meaning when I run a Blitz cache clear (off the back of an entry save) I need to clear the local file cache on each server.

### Describe the solution you would like
It would be great if we had some way to clear the file cache on each server. Perhaps if the cache clear was queued as jobs (one for each server instance, IP's provided in a config file) and then each clear job only runs on the specific server it was dispatched for.

### Describe alternatives you have considered
We are using the Yii storage driver currently (with Redis) but would love to return to the pre-load balancing speeds we experienced with FileStorage. Redis ElastiCache clusters are also pretty pricey when you run the number of sites we do.

I have considered building a plugin for this myself, however it looks as though the cache clear process in Blitz wouldn't support it.

**Comment by bencroker** [comment]
The idea with load balanced servers, as far as I understand them, is that new servers can be spun up or down depending on web traffic. Trying to clear the cache on each load balanced server sounds like a recipe for disaster and is one of the reason that cache stores such as Redis exist. 

If I can help make it easier to hook into the cache clearing process with an event then let me know.

**Comment by mattsbanner** [comment]
Scaling is certainly something that load balancing makes a lot easier. Currently we are load balancing to move away from a single point of failure (e.g if one of our instances goes bad, there are others to carry on serving traffic).

A tool like Varnish caches responses in memory on each node, rather than calling back to a centralised cache store and so I suppose this isn't too out of the ordinary, but is certainly stretching Craft/Yii.

I will take a look at exactly what events we require to add this functionality and shall come back to you.

Thank you for the support!

**Comment by bencroker** [comment]
Varnish certainly has its advantages, but it makes cache purging and integration with Craft trickier. I guess it's a compromise between performance and scalability.

There are plenty of Redis (and ElastiCache) alternatives, which may be worth looking into – but Redis is built to be fast.

**Comment by rostockahoi** [comment]
Hi, I am facing the same challenge using the file cache storage. @mattsbanner Did you come up with a solution?

I guess there is no Redis cache storage implementation as of now? 🙏

**Comment by mattsbanner** [comment]
Hi @rostockahoi. I've had to put this off for the moment, currently we're just continuing to pay for the ElastiCache clusters.

I am still keen to get file storage working on our load balanced servers for the speed improvements and load reduction on PHP. Not to mention the ElastiCache savings as our sites are fairly sizeable and consume memory quite quickly.

Just to confirm, there is a Redis cache storage implementation. This issue is for using the file cache storage but across multiple web server instances. To use Redis you just need to adjust the [cache driver config setting](https://github.com/putyourlightson/craft-blitz/blob/5ed27ca5d0faa81889d1dccd18d59fd4fbd06ece/src/config.php#L48-L49) to use `putyourlightson\blitz\drivers\storage\YiiCacheStorage`.

Then, you can define a Redis cache driver within `config/app.php`.

```php
'components' => [
    'redis' => [
        'class' => yii\redis\Connection::class,
        'hostname' => App::env('REDIS_HOST'),
        'port' => App::env('REDIS_PORT') ?: 6379,
        'password' => App::env('REDIS_PASSWORD') ?: null,
    ],
    'cache' => [
        'class' => yii\redis\Cache::class,
        'defaultDuration' => 86400,
        'keyPrefix' => App::env('APP_ID') ?: 'CraftCMS',
        'enableReplicas' => true,
        'replicas' => [
            'redis',
            ['hostname' => App::env('REDIS_REPLICA_HOST')]
        ]
    ]
]
```

**Comment by teekay** [comment]
@bencroker would you mind sharing your reasoning behind your point that trying to clear the file cache on multiple servers running behind a load balancer would be a "recipe for disaster"?

It might very well be, I would just love to hear your argument. The thing is, just like the original poster I love the performance of a file-based cache.

**Comment by bencroker** [comment]
@teekay The recipe for disaster reasoning is that load balancers come in so many flavours (hardware/software/virtual, network/HTTP/internal, etc.). Trying to make Blitz work with every possible setup is unfeasible.

Generally speaking, if the file system is ephemeral, as it often is in a load balanced environment, then a distributed memory based storage system such as Redis is recommended. All load balanced servers can then access cached pages, rather than only those servers that have already cached the page. As for cache purging, there is only one system to purge, rather than each individual server.

I could be totally wrong about this for your specific server setup, so please let me know if I am. And I would love to see someone tackle this with a custom module/plugin. I'm also happy to add events if and wherever they are needed.

**Comment by hiasl** [comment]
@bencroker & Co: I have exactly the same issue at the moment. would it be ok to put the cache directory on an NFS?

**Comment by bencroker** [comment]
@hiasl It probably would be, but why not use Redis, which is designed for this purpose exactly, instead?

**Comment by hiasl** [comment]
... it's complicated (not Redis - the situation) ...

short version: we have NFS ready, but not Redis

the long version: we currently have 3 single points of failure: 
1.) Load Balancer 
2.) Database 
3.) NFS. 
If we used REDIS we add another SPOF, which is not wanted by our customer, because they do not provide it in their datacenter yet (LB, DB, NFS are provided and maintained).

We could, of course, use Redis in a cluster on each webserver node, but that would require at least 3 nodes and we only have 2. As far as I know Master/Master replication with Redis on 2 nodes is not supported by Redis, so we consider switching to KeyDB, which supports that, but this is not ready yet and will take some more month.

**Comment by bencroker** [comment]
Well, provided Craft can write to the NFS file system (the path is customisable via [config settings](https://github.com/putyourlightson/craft-blitz/blob/8dc32b6711770c97fca48fe2b3a6b40fee6fdcef/src/config.php#L51-L56)) then it should work.

**Comment by teekay** [comment]
> 
> 
> The idea with load balanced servers, as far as I understand them, is that new servers can be spun up or down depending on web traffic. Trying to clear the cache on each load balanced server sounds like a recipe for disaster and is one of the reason that cache stores such as Redis exist.
> 
> If I can help make it easier to hook into the cache clearing process with an event then let me know.

@bencroker would it be feasible for you to enable hooking into the cache clearing process after all, please, or is there a way already? We're looking into this and will try the file-based cache with a static number of Heroku dynos in a Private Space that we can let communicate with one another. Redis works OK but the file cache is faster by a several orders of magnitude. Thanks!

**Comment by bencroker** [comment]
Sure, just added `beforeDeleteUris`, `afterDeleteUris`, `beforeDeleteAll` and `afterDeleteAll` events in https://github.com/putyourlightson/craft-blitz/commit/892417b1194b6c1f774c07715082ca7396b3f5ec for the next release.

**Comment by bencroker** [comment]
FYI these events were released in version 3.10.5.

**Comment by bencroker** [comment]
@teekay or anyone who ends us implementing this, feel free to share your solution and I can see if we can make this either part of the core plugin or a first-party package that we would support.
